File: 1727-largest-submatrix-with-rearrangements/1727-largest-submatrix-with-rearrangements.cpp
using vi = vector<int>;
using vvi = vector<vi>;
class Solution {
public:
  int largestSubmatrix(vector<vector<int>>& a) {
    int n = a.size();
    int m = a[0].size();
    vvi b(m, vi(n));
    for (int j=0; j<m; j++) {
      for (int i=n-1; i>=0; i--) {
        if (i==n-1) b[j][i] = a[i][j];
        else {
          if (a[i][j]) b[j][i] = b[j][i+1] + 1;
        }
      }
    }
    int ans = 0;
    for (int i=0; i<n; i++) {
      vi c;
      for (int j=0; j<m; j++) {
        c.push_back(b[j][i]);
      }
      sort(c.rbegin(), c.rend());
      int h = 1e9;
      for (int j=0; j<m; j++) {
        h = min(h, c[j]);
        ans = max(ans, h * (j+1));
      }
    }
    return ans;
  }
};

File: 1727-largest-submatrix-with-rearrangements/1727-largest-submatrix-with-rearrangements.kt
class Solution {
    fun largestSubmatrix(a: Array<IntArray>): Int {
        val n = a.size
        val m = a[0].size
        var ans = 0
        val b = Array(n) { IntArray(m) }

        for (j in 0 until m) {
            var cnt = 0
            for (i in n - 1 downTo 0) {
                if (a[i][j]== 1) b[i][j] = ++cnt
                else {
                    b[i][j] = 0
                    cnt = 0
                }
            }
        }

        for (i in 0 until n) {
            val t = b[i].toMutableList()
            t.sortDescending()
            for (j in 0 until m) {
                ans = maxOf(ans, (j + 1) * t[j])
            }
        }
        return ans
    }
}
